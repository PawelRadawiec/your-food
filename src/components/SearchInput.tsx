import React, { useEffect, useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { Input } from 'react-native-elements';
import { BallIndicator } from 'react-native-indicators';
import { debounceTime, Subject } from 'rxjs';

const SearchInput = ({
  onSearch,
  loading,
}: {
  onSearch: any;
  loading: boolean;
}) => {
  const [phrase, setPhrase] = useState('');
  const [searchSubject] = useState(new Subject<string>());
  useEffect(() => {
    const subscription = searchSubject
      .pipe(debounceTime(1000))
      .subscribe((value: string) => {
        onSearch(value);
      });
    return () => {
      subscription.unsubscribe();
    };
  }, []);
  return (
    <View style={styles.container}>
      <Input
        placeholder="Search"
        value={phrase}
        disabled={loading}
        onChangeText={(value: string) => {
          setPhrase(value);
          searchSubject.next(value);
        }}
      />
      {loading ? <BallIndicator style={styles.spinner} size={20} /> : null}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    width: 300,
  },
  spinner: {
    marginBottom: 20,
    marginLeft: 10,
  },
});

export default SearchInput;
