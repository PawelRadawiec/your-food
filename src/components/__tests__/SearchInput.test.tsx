import * as React from 'react';
import { Input } from 'react-native-elements';
import renderer, { act, ReactTestInstance } from 'react-test-renderer';
import SearchInput from '../SearchInput';

describe('search input', () => {
  let component: ReactTestInstance;
  let onSearchMock: any;
  beforeAll(() => {
    onSearchMock = jest.fn(() => {});
    component = renderer.create(
      <SearchInput onSearch={onSearchMock} loading={false} />
    ).root;
  });
  it(`renders correctly`, () => {
    const tree = renderer
      .create(<SearchInput onSearch={() => {}} loading={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shoule have one input', () => {
    const root: ReactTestInstance = renderer.create(
      <SearchInput onSearch={() => {}} loading={false} />
    ).root;
    const inputElements = root.findAllByType(Input);
    expect(inputElements.length).toBe(1);
  });

  it('should call onSearch after 1s', async () => {
    const searchInput = component.findByType(Input);
    act(() => searchInput.props.onChangeText('Value'));
    await new Promise((r) => setTimeout(r, 1000));
    expect(onSearchMock.mock.calls.length).toBe(1);
  });
});
