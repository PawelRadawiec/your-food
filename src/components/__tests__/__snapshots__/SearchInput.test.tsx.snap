// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search input renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "width": 300,
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        editable={true}
        onChangeText={[Function]}
        placeholder="Search"
        placeholderTextColor="#86939e"
        rejectResponderTermination={true}
        style={
          Object {
            "color": "#242424",
            "flex": 1,
            "fontSize": 18,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Text
      style={
        Object {
          "color": "#ff190c",
          "fontSize": 12,
          "margin": 5,
        }
      }
    />
  </View>
</View>
`;
